### ISSUE TITLE
Allow FxA to update user's email address without the using logging in

### ISSUE THREAD

**Issue by eviljeff** [mozilla/addons-server#6451]
FxA allows a second email address to be added to an FxA account, and will soon allow that second email address to be made the primary, effectively allowing the user to change their email address. (See https://bugzilla.mozilla.org/show_bug.cgi?id=1393588 for more background.)

AMO updates the email address, if changed, on login, so we will get the new email next time they log in to AMO.  But in the meantime AMO has the old email address so we may then be emailing incorrectly about editor reviewers, user reviews, admin matters, etc.

Question: could one of the existing API endpoints be trivially used to allow this?  /authenticate  and /login both do the update.  But also both mark the user as having logged in.  (And are there additional state and tokens that FxA wouldn't have?)

Otherwise, we need a special endpoint just for FxA.

**Comment by eviljeff** [comment]
@diox @EnTeQuAk ( @mstriemer )

**Comment by rfk** [comment]
> But in the meantime AMO has the old email address so we may then be emailing
> incorrectly about editor reviewers, user reviews, admin matters, etc.

Another option may be to "pull" the latest email address from FxA when you need to send the user an email, using the OAuth access token(s) you got when the user logged in.

Of course, I've no idea how frequently AMO sends email so this may be totally impractical in practice, but it's worth having all the options on the table...

**Comment by eviljeff** [comment]
AMO sends on average about 10,000 emails a day (according to ops looking at socketlab's dashboard).  We don't have a breakdown where they all go (I'd imagine at least a few hundred go to our mailing lists so don't relate to an account's email address) but that's a lot of pulling from FxA.   We could potentially optimize it by only pulling every day or whatever but it still seems we'd be pinging FxA orders of magnitude more times than users would be changing their email address.

**Comment by rfk** [comment]
Ack.  Sounds like "push" is going to be better than "pull" then :-)

Two broad options I see:

1.  We add some back-end process to AMO that can pull notifications off an SQS queue.  We already send SQS notifications to other services (sync, marketing) when a user changes their FxA profile data, so it'd be fairly easy for us to plug AMO into this as well.

2. We add some sort of "webhook" capability and have FxA ping an endpoint on AMO when the user's profile data changes.  That's like to be more dev work on our side and hence slower to get up and running, but may be preferred if you want to avoid tight coupling FxA and AMO infra.

**Comment by rfk** [comment]
(Adding a dedicated webhook endpoint seems preferable than having us do some sort of fake login on an existing endpoint, from a "less likely for bugs to cause security problems" standpoint)

**Comment by muffinresearch** [comment]
@rfk the option for the SQS queue sounds good, especially if that potentially gives us more information when users change things in FXA.

**Comment by eviljeff** [comment]
@rfk how much work is there on the FxA side for us to hook into the SQS queue?  And can you point us to how Sync/Marketing have implemented their end for pulling the notifications?

**Comment by rfk** [comment]
> how much work is there on the FxA side for us to hook into the SQS queue?

Pretty minimal, assuming you're also running in AWS.  We just recently went through this with the marketing team so we have a good idea of how to do it.  Example queue-handling code is at:

  https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/scripts/process_account_events.py

Which in this case doesn't have to handle the change-email event, but should give you the general idea.

**Comment by eviljeff** [comment]
@rfk can you confirm the format of the message we'll be getting, along with the queue name and aws region.

**Comment by eviljeff** [comment]
I started hacking together something based on example, with some changes (e.g. AMO uses django settings, rather than a config file)

This seems like a good thing to be developed into a client library (hint hint 😉)

**Comment by rfk** [comment]
>can you confirm the format of the message we'll be getting

Because of the various wrappings for transport through SQS and SNS, you should receive a message like this:

```
'{"Message": "{\\"email\\": \\"new-primary-email@example.com\\", \\"event\\": \\"primaryEmailChanged\\", \\"uid\\": \\"ABCDEF012345689\\"}"}'
```

That is, a JSONified dict with a "Message" key, whose value is a JSONified dict with event type, affeted uid, and the new primary email.

> along with the queue name and aws region

@jbuck will provision these and let you know the details.  @jbuck should we open a tracking bug for this in bugzilla?

**Comment by rfk** [comment]
Hmmm...it also occurs to me that SQS does not guarantee order of message delivery.  So if a user changes their primary email to "test1@example.com" and then shortly after changes it again to "test2@example.com", it would be possible for you to receive the notification message for "test2@example.com" before the message for "test1@example.com".

We could avoid ordering issues here by having AMO make a request to FxA to fetch latest email address, rather than just using the one contained in the notification.  Does AMO store FxA OAuth tokens for its users, that might allow it to fetch the updated profile information?

**Comment by eviljeff** [comment]
> We could avoid ordering issues here by having AMO make a request to FxA to fetch latest email address, rather than just using the one contained in the notification. Does AMO store FxA OAuth tokens for its users, that might allow it to fetch the updated profile information?

I don't _think_ we do that (from a security perspective, is that a wise thing to do anyway?).  TBH having to do a callback seems rather messy, and kinda defeating the purpose of notification queues - can you just put an event timestamp in the message?

**Comment by rfk** [comment]
>  can you just put an event timestamp in the message?

Yes, if you're happy to remember it on your side and discard notifications with timestamp < the last change you already processed.  I don't enjoy using timestamps because they always mean the possibility for race conditions etc, but the probability of issues there is likely vanishingly small in practice in this case.

**Comment by eviljeff** [comment]
I think timestamps are preferable in this case.  Or any other way of indicating sequence - a unique numeric event identifier (assuming it was guaranteed to be greater than previous events)

**Comment by rfk** [comment]
I filed a bug to add this on our side: https://github.com/mozilla/fxa-auth-server/issues/2160

**Comment by eviljeff** [comment]
I did https://github.com/eviljeff/olympia/commit/b91a84074f85b994aae5db08306aedd8fe876734 (needs adjusting based on the timestamp format you settle on, etc) which would be great to try out with some real-life data once you have the sqs stuff set up for us.

**Comment by jbuck** [comment]
@eviljeff I've setup the change queues in dev & stage, the ARNs are arn:aws:sqs:us-east-1:927034868273:amo-account-change-dev and arn:aws:sqs:us-east-1:142069644989:amo-account-change-stage

**Comment by ValentinaPC** [comment]
Please add some STR for us to check this, if possible, or the label "qa: not needed".
Thanks!

**Comment by eviljeff** [comment]
It will need QA but I need to have ops set something up first to have it the command run automatically.  I'll update once it's testable.

**Comment by eviljeff** [comment]
ops bug https://bugzilla.mozilla.org/show_bug.cgi?id=1419516

**Comment by bqbn** [comment]
Is the AWS SQS ever going to have too many messages at a time?

I'm asking because I don't know if AMO side of code is rate limited, and if it is not, is it a concern that it injects too many tasks into our system and somehow overloads it?

**Comment by muffinresearch** [comment]
@eviljeff @bqbn is there an update on where were have got to with this so far? Are there any open questions?

**Comment by bqbn** [comment]
My latest update is here: https://bugzilla.mozilla.org/show_bug.cgi?id=1419516

tldr; it's deployed on -dev and -stage, waiting to be tested and a go-ahead signal to be deployed to -prod.

**Comment by muffinresearch** [comment]
Thanks @bqbn 

@eviljeff would be good to get this QA'd and then we can look to turn this on in prod once we're happy everything looks good. Could you add some notes for QA to know how and where to test this and what it should look like if it works etc.

**Comment by eviljeff** [comment]
@bqbn 
>Is the AWS SQS ever going to have too many messages at a time?

Messages are retrieved in batches of 10.  For each of them the body of the message is processed, and if it's a `primaryEmailChanged` event the details (email, fxa_id, timestamp) are sent off to a celery task (`primary_email_change_event`) to do the update.  Non-`primaryEmailChanged` events aren't sent to celery, and each message is deleted from the sqs queue regardless.  Then the next 10 is received. 

Theoretically if FxA processes a large number of email change events it would result in a large number of tasks being added to the `default` celery queue.  Though a maximum of 10 at a time, and only `primaryEmailChanged` events. 

I don't know if there are any rate limiting controls implemented on the FxA side, or in SQS directly.

**Comment by jbuck** [comment]
There are no rate-limiting controls implemented on our side, and there are currently 7M messages in the queue in prod. I'd recommend purging that queue before using it

**Comment by eviljeff** [comment]
There's an argument for processing them, to catch any primary email changes we missed... but 7 million is a lot and if the user logged into AMO in the meantime we'd have updated the email then so they'd be obsolete.

On balance, purging is probably for the best and we can look into a manual correction for historic changes as a follow-up.

**Comment by eviljeff** [comment]
QA:
This can be tested on -dev only .  I'll file a follow-up issue to have right settings added for stage&prod so you can do stage/prod QA when that one is closed.

1. Login to AMO with a Firefox Account
2. go to your account profile edit page and load the 'Manage Firefox Account' link in a different tab
3. add a secondary email address in FxA; go through the email confirmation dance; switch secondary email to primary email.
4. flip back to the account profile edit page on AMO and note the email address (it will be the old primary)
5. refresh the page and see the new primary email address.

You don't have to use the Manage link within AMO, or have the edit profile page opened -or be logged into AMO at all- it's just a convenient page to see the change with a refresh.  To verify that:
1. choose an account (A) and make sure you're _not_ logged into AMO
2. change your primary email address with FxA as above
3. login to AMO with a _different_ account (B), that has admin superpowers, in a PrivateBrowsing window and check the email address of account (A) (e.g. using the Manage User button in legacy frontend)

**Comment by AlexandraMoga** [comment]
I've verified both scenarios mentioned above and the primary email update is reflected in AMO:
- by refreshing the Edit profile page opened in a separate window after the primary email is changed from FxA
- by checking the user profile in Admin Tools after the primary email is changed from FxA

I'm marking this issue as verified fixed
